charm-slurm-configurator/src/charm.py

```diff
@@ -289,31 +288,56 @@ def _check_status(self):
         slurmdbd_available = self._stored.slurmdbd_available
         slurmd_available = self._stored.slurmd_available
         slurm_installed = self._stored.slurm_installed
-        default_partition = self._stored.default_partition
-
-        deps = [
-            default_partition,
-            slurmctld_available,
-            slurmdbd_available,
-            slurmd_available,
-            slurm_installed,
-        ]
-
-        if not all(deps):
-            if not slurmctld_available:
-                self.unit.status = BlockedStatus("NEED RELATION TO SLURMCTLD")
-            elif not slurmdbd_available:
-                self.unit.status = BlockedStatus("NEED RELATION TO SLURMDBD")
-            elif not slurmd_available:
-                self.unit.status = BlockedStatus("NEED RELATION TO SLURMD")
-            elif not slurm_installed:
-                self.unit.status = BlockedStatus("SLURM NOT INSTALLED")
+
+        slurmctld_joined = self._slurmctld.is_joined
+        slurmdbd_joined = self._slurmdbd.is_joined
+        slurmd_joined = self._slurmd.is_joined
+
+        relations_needed = []
+        waiting_on = []
+
+        msg = str()
+
+        if not slurmctld_available:
+            if not slurmctld_joined:
+                relations_needed.append("slurmctld")
+            else:
+                waiting_on.append("slurmctld")
+
+        if not slurmdbd_available:
+            if not slurmdbd_joined:
+                relations_needed.append("slurmdbd")
+            else:
+                waiting_on.append("slurmdbd")
+
+        if not slurmd_available:
+            if not slurmd_joined:
+                relations_needed.append("slurmd")
             else:
-                self.unit.status = BlockedStatus("PARTITION NAME UNAVAILABLE")
-            return False
+                waiting_on.append("slurmd")
+
+        if not slurm_installed:
+            waiting_on.append("snap install")
+
+        relations_needed_len = len(relations_needed)
+        waiting_on_len = len(waiting_on)
+
+        if relations_needed_len > 0:
+            msg += f"Needed relations: {','.join(relations_needed)} "
+
+        if waiting_on_len > 0:
+            msg += f"Waiting on: {','.join(waiting_on)}"
+
+        # Using what we have gathered about the status of each slurm component,
+        # determine the application status.
+        if relations_needed_len > 0:
+            self.unit.status = BlockedStatus(msg)
+        elif waiting_on_len > 0:
+            self.unit.status = WaitingStatus(msg)
         else:
             self.unit.status = ActiveStatus("")
```

Comment:
do we want to set it to Available instead of blank?

Reply:
👍